rct229/rule_engine/rule_base.py

```diff
@@ -113,26 +114,33 @@ def evaluate(self, rmrs, data=None):
                     # Get calculated values; these can be used by
                     # manual_check_required() or rule_check() and will
                     # be included in the output
-                    calc_vals = self.get_calc_vals(context, data)
-                    if calc_vals is not None:
-                        outcome["calc_vals"] = calc_vals
-
-                    # Determine if manual check is required
-                    if self.manual_check_required(context, calc_vals, data):
-                        outcome["result"] = "MANUAL_CHECK_REQUIRED"
-                    else:
-                        # Evaluate the actual rule check
-                        result = self.rule_check(context, calc_vals, data)
-                        if isinstance(result, list):
-                            # The result is a list of outcomes
-                            outcome["result"] = result
-                        # Assume result type is bool
-                        elif result:
-                            outcome["result"] = "PASSED"
+                    try:
+                        calc_vals = self.get_calc_vals(context, data)
+                        if calc_vals is not None:
+                            outcome["calc_vals"] = calc_vals
+
+                        # Determine if manual check is required
+                        if self.manual_check_required(context, calc_vals, data):
+                            outcome["result"] = "UNDETERMINED"
                         else:
-                            outcome["result"] = "FAILED"
+                            # Evaluate the actual rule check
+                            result = self.rule_check(context, calc_vals, data)
+                            if isinstance(result, list):
+                                # The result is a list of outcomes
+                                outcome["result"] = result
+                            # Assume result type is bool
+                            elif result:
+                                outcome["result"] = "PASSED"
+                            else:
+                                outcome["result"] = "FAILED"
+                    except MissingKeyException as ke:
+                        outcome["result"] = "UNDETERMINED"
+                        outcome["message"] = str(ke)
+                    except RCTFailureException as fe:
+                        outcome["result"] = "FAILED"
+                        outcome["message"] = str(fe)
                 else:
-                    outcome["result"] = "NA"
+                    outcome["result"] = "UNDETERMINED"
```

Comment:
Do we really want to make this UNDETERMINED?  It seems that NA is distinctly different. I suggest leaving is a NA until a further discussion with Supriya about it.